compiler/rustc_lint/src/non_local_def.rs

```diff
@@ -155,9 +169,54 @@ impl<'tcx> LateLintPass<'tcx> for NonLocalDefinitions {
                     .map(|of_trait| path_has_local_parent(of_trait.path, cx, parent, parent_parent))
                     .unwrap_or(false);
 
-                // If none of them have a local parent (LOGICAL NOR) this means that
-                // this impl definition is a non-local definition and so we lint on it.
-                if !(self_ty_has_local_parent || of_trait_has_local_parent) {
+                // Detecting if the impl definition is leaking outside of it's defining scope.
+                //
+                // Rule: for each impl, instantiate all local types with inference vars and
+                // then assemble candidates for that goal, if there are more than 1 (non-private
+                // impls), it does not leak.
+                //
+                // https://github.com/rust-lang/rust/issues/121621#issuecomment-1976826895
+                let impl_is_not_leaky = cx
+                    .tcx
+                    .impl_trait_ref(def_id)
+                    .map(|binder| {
+                        let infcx = cx.tcx.infer_ctxt().build();
+                        let trait_ref = binder
+                            .instantiate(cx.tcx, infcx.fresh_args_for_item(item.span, def_id));
+
+                        let trait_ref = trait_ref.fold_with(&mut LocalTypeInferenceFolder {
+                            infcx: &infcx,
+                            to_ignore: 1,
+                            impl_parent: parent,
+                            impl_parent_parent: parent_parent,
+                            span: item.span,
+                        });
+
+                        let poly_trait_obligation = Obligation::new(
+                            cx.tcx,
+                            ObligationCause::dummy(),
+                            ty::ParamEnv::empty(),
+                            Binder::dummy(trait_ref),
```

Comment:
```suggestion
                            trait_ref,
```
`Binder::dummy` is unnecessary

Reply:
According to compiler it's not unnecessary.

```text
error[E0277]: the trait bound `rustc_middle::ty::TraitRef<'_>: ToPredicate<'_, rustc_middle::ty::Binder<'_, TraitPredicate<'_>>>` is not satisfied
   --> compiler/rustc_lint/src/non_local_def.rs:198:29
    |
194 |                         let poly_trait_obligation = Obligation::new(
    |                                                     --------------- required by a bound introduced by this call
...
198 |                             trait_ref,
    |                             ^^^^^^^^^ the trait `ToPredicate<'_, rustc_middle::ty::Binder<'_, TraitPredicate<'_>>>` is not implemented for `rustc_middle::ty::TraitRef<'_>`
    |
    = help: the following other types implement trait `ToPredicate<'tcx, P>`:
              <rustc_middle::ty::TraitRef<'tcx> as ToPredicate<'tcx>>
              <rustc_middle::ty::TraitRef<'tcx> as ToPredicate<'tcx, TraitPredicate<'tcx>>>
              <rustc_middle::ty::TraitRef<'tcx> as ToPredicate<'tcx, rustc_middle::ty::Clause<'tcx>>>
note: required by a bound in `Obligation::<'tcx, O>::new`
   --> /home/archie/Projects/RustProjects/rust/compiler/rustc_infer/src/traits/mod.rs:158:25
    |
154 |     pub fn new(
    |            --- required by a bound in this associated function
...
158 |         predicate: impl ToPredicate<'tcx, O>,
    |                         ^^^^^^^^^^^^^^^^^^^^ required by this bound in `Obligation::<'tcx, O>::new`
```